conditional_camera/conditional_cam.go

```diff
@@ -0,0 +1,253 @@
+			if err != nil {
+				return nil, err
+			}
+
+			return fc, nil
+		},
+	})
+}
+
+type cachedData struct {
+	imgs []camera.NamedImage
+	meta resource.ResponseMetadata
+}
+
+type conditionalCamera struct {
+	resource.AlwaysRebuild
+	resource.TriviallyCloseable
+
+	name   resource.Name
+	conf   *Config
+	logger logging.Logger
+
+	cam     camera.Camera
+	filtSvc resource.Resource
+
+	mu          sync.Mutex
+	buffer      []cachedData
+	toSend      []cachedData
+	captureTill time.Time
+}
+
+func (cc *conditionalCamera) Name() resource.Name {
+	return cc.name
+}
+
+func (cc *conditionalCamera) DoCommand(ctx context.Context, cmd map[string]interface{}) (map[string]interface{}, error) {
+	return nil, resource.ErrDoUnimplemented
+}
+
+func (cc *conditionalCamera) Images(ctx context.Context) ([]camera.NamedImage, resource.ResponseMetadata, error) {
+	images, meta, err := cc.cam.Images(ctx)
+	if err != nil {
+		return images, meta, err
+	}
+
+	extra, ok := camera.FromContext(ctx)
+	if !ok || extra[data.FromDMString] != true {
+		return images, meta, nil
+	}
+
+	for range images {
+		shouldSend, err := cc.shouldSend(ctx)
+		if err != nil {
+			return nil, meta, err
+		}
+
+		if shouldSend {
+			return images, meta, nil
+		}
+	}
+
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
+
+	cc.addToBuffer_inlock(images, meta)
+
+	if len(cc.toSend) > 0 {
+		x := cc.toSend[0]
+		cc.toSend = cc.toSend[1:]
+		return x.imgs, x.meta, nil
+	}
+
+	return nil, meta, data.ErrNoCaptureToStore
+}
+
+func (cc *conditionalCamera) Stream(ctx context.Context, errHandlers ...gostream.ErrorHandler) (gostream.VideoStream, error) {
+	camStream, err := cc.cam.Stream(ctx, errHandlers...)
+	if err != nil {
+		return nil, err
+	}
+
+	return conditionalStream{camStream, cc}, nil
+}
+
+type conditionalStream struct {
+	cameraStream gostream.VideoStream
+	cc           *conditionalCamera
+}
+
+func (cs conditionalStream) Next(ctx context.Context) (image.Image, func(), error) {
+	extra, ok := camera.FromContext(ctx)
+	if !ok || extra[data.FromDMString] != true {
+		// If not data management collector, return underlying stream contents without filtering.
+		return cs.cameraStream.Next(ctx)
+	}
+
+	img, release, err := cs.cameraStream.Next(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	should, err := cs.cc.shouldSend(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if should {
+		return img, release, nil
+	}
+
+	cs.cc.mu.Lock()
+	defer cs.cc.mu.Unlock()
+
+	cs.cc.addToBuffer_inlock([]camera.NamedImage{{img, ""}}, resource.ResponseMetadata{CapturedAt: time.Now()})
+
+	return nil, nil, data.ErrNoCaptureToStore
+}
+
+func (cc *conditionalCamera) addToBuffer_inlock(imgs []camera.NamedImage, meta resource.ResponseMetadata) {
+	if cc.conf.WindowSeconds == 0 {
+		return
+	}
+
+	cc.cleanBuffer_inlock()
+	cc.buffer = append(cc.buffer, cachedData{imgs, meta})
+}
+
+func (cs conditionalStream) Close(ctx context.Context) error {
+	return cs.cameraStream.Close(ctx)
+}
```

Comment:
So, normally I would say --a lot of this code is copied from the original filtered-camera code. Rather than copy-pasting, you should have both inherit the logic from a common conditionalStream object. BUT, we are going to do a re-factor of this code soon and upgrade the RDK version to the latest version. Which will get rid of all of these stream objects. So you've lucked out this time!